### ISSUE TITLE
add faster iterative moving window method for obtaining distance to roads.

### ISSUE THREAD

**Issue by josie-hughes** [LandSciTech/roads#15]
see https://github.com/LandSciTech/RoadAnalysis/blob/master/scripts/getDistanceFromSource.R

**Comment by see24** [comment]
@josie-hughes I added this in the rast2Lines branch. I think we should use the `terra::focal` function though. In a test case with the 100m by 100m cost raster for Revelstoke from the roads paper the pfocal version required up to 8 GB of memory to be allocated while the terra version only needed 319.2KB. See this test file for the comparison https://github.com/LandSciTech/roads/blob/6d0a28b8aa6a67b9b3f1dbcb8f2a95003add84e9/tests/testthat/test-getDistFromSource.R

Does that make sense to you? In general it is a bit of a weakness for pfocal that it requires a matrix since they can take up a lot of space compared to a raster.

**Comment by see24** [comment]
@josie-hughes one more question for you whenever you have time. 

For describing the `kwidth` and `disagg` arguments to `getDistFromSource` it doesn't seem like the way you described it in the link above holds out. Or I am not understanding it.

This is what the results look like with a few different versions of the options:
![image](https://user-images.githubusercontent.com/39161074/171945717-11d7ce79-3228-472e-90ef-4e12a2d6cece.png)

I have described this behaviour like this 
"There is a tradeoff between speed and accuracy in this method, controlled by the parameters kwidth and dissag. kwidth is the radius of the moving window in number of cells. Smaller values will provide a smoother gradient while larger values will give a blockier result. If dissag is false, the resulting distance map will have values of 0, 1kwidth, 2kwidth, etc. If dissag is true the algorithm disaggregates the raster to a finer resolution, so the resulting distance map will have values of 0,1,2, etc. Setting dissag = TRUE or a low value for kwidth will increase the computation time. " 

Does that seem accurate and does the behaviour in the picture look right to you?

**Comment by see24** [comment]
Have the option for pfocal. Maybe as not the default since could mess up some users but we want access

**Comment by see24** [comment]
Can't see effent of kwidth as well without wiggly roads. Josie says to just remove the disagg option. 
Write a vignette about this with disagg because without it you can't get an answer that is finer than 3 pixels 

disagg is essentially breaking the raster into smaller pieces first so that you can get less grid artifacts and fine scale distances.
Use raster with histogram to show what's wrong with some Show copute time, hist and looking at the map blockiness

**Comment by see24** [comment]
@josie-hughes I added a new vignette for the getDistanceFromRoads function. I also added another method for getting the distance from "global" moving window rather than an iterative moving window. This version can be slow for large maxDists but was much faster in many cases. Here is a new picture comparing the methods.

https://github.com/LandSciTech/roads/blob/9c77ca46cfaa1a641935c17354bddf670d5e04f6/vignettes/articles/getDistFromSource_files/figure-gfm/make_plots-1.png

And the vignette is here https://github.com/LandSciTech/roads/blob/9c77ca46cfaa1a641935c17354bddf670d5e04f6/vignettes/articles/getDistFromSource.Rmd

Should still add some more tests for behavior with NAs and other situations
